src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPServerSideOnTypeFormattingHelper.java

```diff
@@ -0,0 +1,151 @@
+        String charTypedAsString = String.valueOf(charTyped);
+        if (!triggerCharacters.contains(charTypedAsString)) {
+            return false;
+        }
+
+        // If so, issue a request for on-type formatting
+        int offset = editor.getCaretModel().getOffset();
+        Document document = editor.getDocument();
+        Position position = LSPIJUtils.toPosition(offset, document);
+        DocumentOnTypeFormattingParams onTypeFormattingParams = new DocumentOnTypeFormattingParams(
+                LSPIJUtils.toTextDocumentIdentifier(file.getVirtualFile()),
+                new FormattingOptions(LSPIJUtils.getTabSize(editor), LSPIJUtils.isInsertSpaces(editor)),
+                position,
+                charTypedAsString
+        );
+        CompletableFuture<List<TextEdit>> onTypeFormattingFutures = onTypeFormattingSupport.onTypeFormatting(onTypeFormattingParams);
+        try {
+            waitUntilDone(onTypeFormattingFutures, file);
+        } catch (ProcessCanceledException e) {
+            //Since 2024.2 ProcessCanceledException extends CancellationException so we can't use multicatch to keep backward compatibility
+            //TODO delete block when minimum required version is 2024.2
+            onTypeFormattingSupport.cancel();
+            return false;
+        } catch (CancellationException e) {
+            onTypeFormattingSupport.cancel();
+            return false;
+        } catch (ExecutionException e) {
+            LOGGER.error("Error while consuming LSP 'textDocument/onTypeFormatting' request", e);
+            return false;
+        }
+
+        if (!isDoneNormally(onTypeFormattingFutures)) {
+            return false;
+        }
+
+        // If on-type formatting resulted in text edits, apply them
+        List<TextEdit> textEdits = onTypeFormattingFutures.getNow(Collections.emptyList());
+        if (ContainerUtil.isEmpty(textEdits)) {
+            return false;
+        }
+
+        LSPIJUtils.applyEdits(editor, document, textEdits);
+        return true;
+    }
+
+    /**
+     * Returns the LSP on-type formatting information for the provided file if supported. This includes the on-type
+     * formatting options and the on-type formatting support.
+     *
+     * @param file the PSI file
+     * @return the LSP on-type formatting information for the file if supported; otherwise null
+     */
+    @Nullable
+    private static Pair<DocumentOnTypeFormattingOptions, LSPOnTypeFormattingSupport> getOnTypeFormattingInfo(@NotNull PsiFile file) {
+        // On-type formatting shouldn't trigger a language server to start
+        Project project = file.getProject();
+        Set<LanguageServerWrapper> startedLanguageServers = LanguageServiceAccessor.getInstance(project).getStartedServers();
+        for (LanguageServerWrapper startedLanguageServer : startedLanguageServers) {
+            // TODO: Is there a better way to ask if this started language server supports the file?
```

Comment:
This is the same as what I have in the client-side on-type formatting handler in the other PR/branch, so the same question holds. We don't want to suspend the IDE in response to the user typing a character to start a language server, so we first check whether there is a started language server for the current file. Is that the best way to do this? If not, what should I be doing instead?

Reply:
Nice catch! I think you can do that with the standard getLanguageServers by using a before filter which matches ybe server status with started. 

In the filter you receive the languageserver item which defines a getServerStatus if I remember

Reply:
Okay. Do you think it's worth switching to that? Is there anything wrong with the approach taken here?